torchvision/transforms/v2/functional/_utils.py

```diff
@@ -1,8 +1,110 @@
+
+    # this avoids converting list -> set at runtime below
+    kernel_params = set(kernel_params)
+
+    @functools.wraps(kernel)
+    def wrapper(inpt, *args, **kwargs):
+        input_type = type(inpt)
+
+        if needs_args_kwargs_handling:
+            # Convert args to kwargs to simplify further processing
+            kwargs.update(dict(zip(dispatcher_params, args)))
+            args = ()
+
+            # drop parameters that are not relevant for the kernel, but have a default value
+            # in the dispatcher
+            for kwarg in kwargs.keys() - kernel_params:
+                del kwargs[kwarg]
+
+            # add parameters that are passed implicitly to the dispatcher as metadata,
+            # but have to be explicit for the kernel
+            for kwarg in input_type.__annotations__.keys() & kernel_params:
+                kwargs[kwarg] = getattr(inpt, kwarg)
+
+        output = kernel(inpt.as_subclass(torch.Tensor), *args, **kwargs)
+
+        if isinstance(inpt, datapoints.BoundingBox) and isinstance(output, tuple):
+            output, spatial_size = output
+            metadata = dict(spatial_size=spatial_size)
+        else:
+            metadata = dict()
+
+        return input_type.wrap_like(inpt, output, **metadata)
+
+    return wrapper
+
+
+def _register_kernel_internal(dispatcher, datapoint_cls, *, wrap_kernel=True):
+    registry = _KERNEL_REGISTRY.setdefault(dispatcher, {})
+    if datapoint_cls in registry:
+        raise TypeError(
+            f"Dispatcher '{dispatcher.__name__}' already has a kernel registered for type '{datapoint_cls.__name__}'."
+        )
+
+    def decorator(kernel):
+        registry[datapoint_cls] = _kernel_wrapper_internal(dispatcher, kernel) if wrap_kernel else kernel
+        return kernel
+
+    return decorator
+
+
+def register_kernel(dispatcher, datapoint_cls):
+    return _register_kernel_internal(dispatcher, datapoint_cls, wrap_kernel=False)
+
+
+def _noop(inpt, *args, **kwargs):
+    return inpt
+
+
+def _register_explicit_noops(dispatcher, *datapoints_clss):
+    """
+    Although this looks redundant with the no-op behavior of _get_kernel, this explicit registration prevents users
+    from registering kernels for builtin datapoints on builtin dispatchers that rely on the no-op behavior.
+
+    For example, without explicit no-op registration the following would be valid user code:
+
+    .. code::
+        from torchvision.transforms.v2 import functional as F
+
+        @F.register_kernel(F.adjust_brightness, datapoints.BoundingBox)
+        def lol(...):
+            ...
+    """
+    for datapoint_cls in datapoints_clss:
+        register_kernel(dispatcher, datapoint_cls)(_noop)
+
+
+def _get_kernel(dispatcher, datapoint_cls):
+    registry = _KERNEL_REGISTRY.get(dispatcher)
+    if not registry:
+        raise ValueError(f"No kernel registered for dispatcher '{dispatcher.__name__}'.")
+
+    if datapoint_cls in registry:
+        return registry[datapoint_cls]
```

Comment:
is this already covered by the `for` loop below? (No strong opinion)

Reply:
Since our stuff is the first thing to be registered, you are correct. I thought I make it explicit that an exact type match always beats the subclass check regardless of registration order.

Imagine the following

```python
class Foo(datapoints.Datapoint):
    ...

class Bar(Foo):
    ...

def dispatcher(inpt):
    ...

@F.register_kernel(dispatcher, Foo)
def foo_kernel(foo):
    ...

@F.register_kernel(dispatcher, Bar)
def bar_kernel(bar):
    ...

dispatcher(Foo(...))  # calls foo_kernel, ok
dispatcher(Bar(...))  # calls foo_kernel, oops
```

This scenario is unlikely for now. However, who knows how our stuff is adopted in the future.